File: 0031-next-permutation/0031-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n=nums.size();
        int id=-1;
        for(int i=n-2;i>=0;i--)
        {
            if(nums[i]<nums[i+1])
            {
                id=i;
                break;
            }
        }
        if(id==-1)
        {
            reverse(nums.begin(),nums.end());
        }
        else
        {
        for(int i=n-1;i>id;i--)
        {
            if(nums[i]>nums[id])
            {
                swap(nums[i],nums[id]);
                break;
            }
        }
        int i=id+1;
        int j=n-1;
        while(i<j)
        {
            swap(nums[i++],nums[j--]);
        }
        }
        
    }
};

File: 0031-next-permutation/0031-next-permutation.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums)
{
     let n=nums.length;
    let id=-1;
    for(let i=n-2;i>=0;i--)
        {
            if(nums[i]<nums[i+1])
                {
                    id=i;
                    break;
                }
        }
    
    for(let i=n-1;i>id;i--)
        {
            if(nums[i]>nums[id])
                {
                    nums[i]=nums[i]^nums[id];
                    nums[id]=nums[id]^nums[i];
                     nums[i]=nums[i]^nums[id];
                    break;
                }
        }
    let i=id+1;
    let j=n-1;
    while(i<j)
        {
             nums[i]=nums[i]^nums[j];
                    nums[j]=nums[j]^nums[i];
                     nums[i]=nums[i]^nums[j];
            i++;
            j--;
        }
};